### DISCUSSION TITLE
DI Scoping in Maui - Multi User Application

### DISCUSSION THREAD

**Discussion by shadowfoxish** [dotnet/maui#16295] [Q&A]
Hi all, 
tl;dr : How can I insert a "Scope" that the application can manage into the default DI container from which all subsequent GetService requests will come through?

I have an application that is used by warehouse folks where multiple people may log in/out of the same instance of the app without closing it. I have written stuff to update services when this happens, but my main problem is just that some of the view models and pages (which are all registered as Transient) are lingering between different logins. I don't know why these transients are being re-used but they are and whatever, it's generally fine as long as the user doesn't change.

I'd like to use DI scopes for this so that each time a user logs out a new DI scope is created and essentially all previous instances of the services go away and the new user gets new instances. 

Flow: App Builder -> Create Scope -> Show Login Page -> User session -> User logs out -> Old scope closed, create new scope -> Show Login Page

I feel like this should be very possible, but I'm not sure how to kind of get in the middle of the ServiceProvider and add a custom implementation that allows me to tell it when to start a new scope.

**Comment by Xyncgas** [comment]
Consider a functional approach your UI is produced using { user }

You can have multiple user logged in, but to switch between user you would recreate the UI, leaving no chance for 'lingering UI' as you mentioned that would potentially be a problem.

You can achieve the result using object oriented programming to create objects that represent the UI you are building, and functional oriented programming to avoid shared mutability, don't worry about overhead because it's not backend it's front end

**Reply by shadowfoxish** [reply]
Making the UI completely functional is more or less what I've done, however the UI depends on the various view model types, and those view models seem to be getting scoped the same as the rest of the UI and reused for some reason. 

Your thoughts are inspiring though, I went looking though and the only UI I create myself using "new" is the login page. I can try to create that using a new scope and see what happens. I'm not sure that other pages created by navigation and the framework itself will use the new scope to resolve new dependencies though.
